### PR TITLE
Support FP8 INC in vLLM

### PR DESCRIPTION
**Pull request by nirda7**
FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

File: vllm/attention/backends/habana_attn.py [hunk 1]
```diff
@@ -137,10 +139,16 @@ def __init__(
...
         max_seq_len: int = 4096,
     ) -> None:
+        super(AttentionImpl, self).__init__()
         self.kv_cache_dtype = kv_cache_dtype
         self.num_heads = num_heads
         self.head_size = head_size
         self.scale = float(scale)
+        self.qk_matmul = Matmul()
+        self.softmax = Softmax()
+        self.av_matmul = Matmul()
```
Comment (dudilester) [inline vllm/attention/backends/habana_attn.py:149]: is the name av_matmul intended? I know we used kv_matmul previously. If we are using av_matmul we should also make sure the mlperf_inference/code/vllm-static/hqt/step-2-postprocess_measure.py util is using the same layer names.

Comment (BacharL) [reply vllm/attention/backends/habana_attn.py:149]: also we should align names with optimum-habana-fork
https://github.com/HabanaAI/optimum-habana-fork/blob/74d5dc8a958666a04cc73a8ba25db0cd65f1a7f4/optimum/habana/transformers/models/llama/modeling_llama.py#L339

Comment (nirda7) [reply vllm/attention/backends/habana_attn.py:149]: yes, the name av_matmul intended and we will change accordingly the names in mlperf repo.

Comment (nirda7) [reply vllm/attention/backends/habana_attn.py:149]: Aligning names with optimum-habana-fork won't be done now as there is still work to do on how to sync these models or even allow the INC measurement within vLLM.

Comment (dudilester) [reply vllm/attention/backends/habana_attn.py:149]: @nirda7 I think there is logic in keeping the names similar in habana attn models, we can consider changing the member names to v_cache and k_cache as they are in OHF.

File: vllm/engine/arg_utils.py [hunk 2]
```diff
@@ -223,11 +229,12 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
...
             '--kv-cache-dtype',
             type=str,
-            choices=['auto', 'fp8', 'fp8_e5m2', 'fp8_e4m3'],
+            choices=['auto', 'fp8', 'fp8_e5m2', 'fp8_e4m3', 'fp8_inc'],
             default=EngineArgs.kv_cache_dtype,
             help='Data type for kv cache storage. If "auto", will use model '
             'data type. CUDA 11.8+ supports fp8 (=fp8_e4m3) and fp8_e5m2. '
-            'ROCm (AMD GPU) supports fp8 (=fp8_e4m3)')
+            'ROCm (AMD GPU) supports fp8 (=fp8_e4m3). '
+            'Intel Gaudi (HPU) also supports fp8 (using fp8_inc).')
```
Comment (dudilester) [inline vllm/engine/arg_utils.py:237]: remove the word "also"

File: vllm/worker/habana_model_runner.py [hunk 3]
```diff
@@ -413,6 +413,9 @@ def __init__(
         self._setup_buckets()
 
     def load_model(self) -> None:
+        if self.model_config.quantization == 'inc':
+            import habana_frameworks.torch.core as htcore
```
[the remark this answers is not in the document]
Comment (nirda7) [reply vllm/worker/habana_model_runner.py:417]: Done

File: vllm/engine/arg_utils.py (hunk 2 above)
Comment (nirda7) [reply vllm/engine/arg_utils.py:237]: ok

File: vllm/attention/backends/habana_attn.py (hunk 1 above)
Comment (nirda7) [reply vllm/attention/backends/habana_attn.py]: ok - names changed

**Review by kzawora-intel — APPROVED** [review]
Approved, but we *need* to revert the manual finish_measurements APIs once the hang issue is resolved.

**Pull request merged**